### PR TITLE
feat: implement uv-style hierarchical configuration system

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Summary

Implements a hierarchical configuration system following uv's architecture with the following precedence order:

1. **CLI-provided config** (highest precedence)
2. **Environment variables** (SERPEN_*)
3. **Project config** (./serpen.toml)
4. **User config** (~/.config/serpen/serpen.toml)  
5. **System config** (/etc/serpen/serpen.toml or %SYSTEMDRIVE%\ProgramData\serpen\serpen.toml)
6. **Default values** (lowest precedence)

### 🏗️ **Hierarchical Configuration Loading**
- XDG-compliant directory structure for user and system configs
- Proper config merging with precedence using custom Combine trait
- Cross-platform support (Linux, macOS, Windows)

### 🌍 **Environment Variable Support** 
All config options now support SERPEN_ prefixed environment variables:
- `SERPEN_SRC`: comma-separated source directories
- `SERPEN_KNOWN_FIRST_PARTY`: comma-separated first-party modules
- `SERPEN_KNOWN_THIRD_PARTY`: comma-separated third-party modules
- `SERPEN_PRESERVE_COMMENTS`: true/false/1/0/yes/no/on/off
- `SERPEN_PRESERVE_TYPE_HINTS`: true/false/1/0/yes/no/on/off
- `SERPEN_TARGET_VERSION`: py38/py39/py310/py311/py312/py313

### 📁 **Config File Locations**
- **System**: `/etc/serpen/serpen.toml` (Linux/macOS) or `%SYSTEMDRIVE%\ProgramData\serpen\serpen.toml` (Windows)
- **User**: `~/.config/serpen/serpen.toml` (XDG-compliant)
- **Project**: `./serpen.toml` (current directory)
- **CLI**: `--config` flag

## Implementation Details

- Added `dirs.rs` module for XDG-compliant directory management
- Added `combine.rs` module with Combine trait for config merging
- Enhanced `config.rs` with hierarchical loading and environment variable parsing
- Comprehensive test coverage for all scenarios
- Backward compatible with existing configuration

## Test Plan

✅ All existing tests pass  
✅ Added comprehensive tests for hierarchical config loading  
✅ Added tests for environment variable parsing  
✅ Added tests for config file precedence  
✅ Added tests for XDG directory structure  

🤖 Generated with [Claude Code](https://claude.ai/code)

## Summary by CodeRabbit

- **New Features**
  - Added hierarchical configuration loading, allowing settings to be merged from system, user, project files, environment variables, and CLI config with clear precedence.
  - Introduced support for configuration via environment variables with optional fields for all config parameters.
  - Added utilities to locate user and system configuration directories and files across platforms.
  - Added a new merging trait to consistently combine configuration values with defined precedence rules.

- **Bug Fixes**
  - Improved error handling and validation in configuration loading, including enhanced error message assertions in tests.

- **Tests**
  - Expanded tests to cover configuration merging, environment variable parsing, file loading, directory detection, and target version parsing.

- **Documentation**
  - Enhanced inline documentation for new configuration and directory utilities.
  - Added detailed Git workflow documentation emphasizing CI usage, review process, and commit standards.
  - Expanded README with comprehensive configuration management guidance, including environment variable overrides and config file locations.

### CODE REVIEW

File: crates/serpen/src/combine.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        }
+    };
+}
+
+impl_combine_or!(String);
+impl_combine_or!(bool);
+impl_combine_or!(PathBuf);
+
+impl<T> Combine for Option<Vec<T>> {
+    /// Combine two vectors by extending the higher precedence vector (`self`) with the lower
```
Comment (tinovyatkin) [inline crates/serpen/src/combine.rs:35]: ✅ **Resolved**: Updated the documentation to clearly explain precedence behavior:

```rust
/// Combine two vectors by extending the higher precedence vector (`self`) with the lower 
/// precedence vector (`other`), placing higher precedence items first.
```

This makes it explicit that `self` has higher precedence and its items come first in the merged result.

File: crates/serpen/src/dirs.rs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    user_config_dir().map(|mut path| {
+        path.push("serpen");
+        path
+    })
+}
+
+#[cfg(not(windows))]
+fn locate_system_config_xdg(value: Option<&str>) -> Option<PathBuf> {
+    // On Linux and macOS, read the `XDG_CONFIG_DIRS` environment variable.
+    let default = "/etc/xdg";
```
Comment (tinovyatkin) [inline crates/serpen/src/dirs.rs:28]: ✅ **Resolved**: Removed the redundant `use std::path::Path;` import since `Path` is already imported at the top of the file. This cleans up the code by eliminating unnecessary duplicate imports.

File: crates/serpen/src/dirs.rs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+/// On Unix-like systems, uses the `XDG_CONFIG_DIRS` environment variable (falling back to
+/// `/etc/xdg/serpen/serpen.toml` if unset or empty) and then `/etc/serpen/serpen.toml`
+///
+/// On Windows, uses `%SYSTEMDRIVE%\ProgramData\serpen\serpen.toml`.
+pub fn system_config_file() -> Option<PathBuf> {
+    #[cfg(windows)]
+    {
+        env::var("SYSTEMDRIVE")
+            .ok()
+            .and_then(|system_drive| locate_system_config_windows(PathBuf::from(system_drive)))
```
Comment (tinovyatkin) [inline crates/serpen/src/dirs.rs:62]: ✅ **Resolved**: Simplified Windows path construction by using `PathBuf::from()` instead of manual string formatting:

```rust
env::var("SYSTEMDRIVE")
    .ok()
    .and_then(|system_drive| locate_system_config_windows(PathBuf::from(system_drive)))
```

This is more idiomatic and leverages PathBuf's built-in path handling capabilities.

File: crates/serpen/src/config.rs [hunk 4]
```diff
@@ -72,31 +210,268 @@ impl Config {
...
+        assert_eq!(combined.target_version, "py39");
+
+        // For collections, higher precedence completely replaces
+        assert_eq!(combined.src, vec![PathBuf::from("src1")]);
+        assert!(combined.known_first_party.contains("module1"));
+        assert!(!combined.known_first_party.contains("module2"));
+    }
+
+    #[test]
+    #[serial_test::serial]
```
Comment (tinovyatkin) [inline crates/serpen/src/config.rs:371]: ✅ **Resolved**: Added `#[serial_test::serial]` annotation to prevent race conditions:

```rust
#[test]
#[serial_test::serial]
fn test_env_config_parsing() {
    // ... test code that modifies environment variables
}
```

The `serial_test` crate ensures these tests run sequentially, preventing flaky failures due to concurrent environment variable access. This is especially important since environment variables are global state that can interfere between parallel test executions.

File: crates/serpen/src/config.rs [hunk 5]
```diff
@@ -41,6 +45,140 @@ impl Default for Config {
     }
 }
 
+impl Combine for Config {
+    fn combine(self, other: Self) -> Self {
+        Self {
+            // For collections, higher precedence (self) completely replaces lower precedence (other)
+            // if self has non-default values, otherwise use other
+            src: if self.src != Config::default().src {
```
Comment (tinovyatkin) [inline crates/serpen/src/config.rs:53]: ✅ **Resolved**: Replaced hardcoded default values with `Config::default()` for maintainability:

```rust
src: if self.src != Config::default().src {
    self.src
} else {
    other.src
},
```

This ensures the comparison always uses the current default value definition, making the code more resilient to future changes in default values. The default is now defined in one place (the `Default` implementation) and referenced consistently.

**Review by tinovyatkin — COMMENTED** [review]
## ✅ All CodeRabbit Review Comments Addressed

I've successfully addressed all 7 CodeRabbit review comments with the following improvements:

### 🔧 **Code Quality Improvements**
1. **Enhanced Documentation** - Clarified vector merging precedence behavior in combine.rs
2. **Code Cleanup** - Removed redundant imports in dirs.rs  
3. **Idiomatic Rust** - Simplified Windows path construction using PathBuf methods
4. **Robust Default Handling** - Replaced hardcoded values with `Config::default()` references

### 🛡️ **Reliability & Safety**
5. **Thread Safety** - Added `#[serial_test::serial]` for environment variable tests to prevent race conditions
6. **Panic Safety** - Implemented RAII guard pattern for directory restoration in tests
7. **Input Validation** - Enhanced comma-separated parsing with proper edge case handling

### 🧪 **Testing Robustness** 
- Tests now run sequentially when modifying global state (environment variables)
- Directory changes are automatically restored even on test panic via Drop trait
- Environment variable parsing handles empty strings, trailing commas, and whitespace

All changes maintain backward compatibility while improving code maintainability and test reliability. The implementation follows Rust best practices and ensures robust error handling throughout.

File: crates/serpen/src/combine.rs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    fn combine(self, other: Option<Vec<T>>) -> Option<Vec<T>> {
+        match (self, other) {
+            (Some(mut a), Some(b)) => {
+                a.extend(b);
+                Some(a)
+            }
+            (a, b) => a.or(b),
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/serpen/src/combine.rs]: This is a test reply
